samza-core/src/main/java/org/apache/samza/coordinator/metadatastore/CoordinatorStreamStore.java

```diff
@@ -188,6 +188,14 @@ public void flush() {
     }
   }
 
+  @Override
+  protected void finalize() throws Throwable {
```

Comment:
I don't think using finalize is a good idea.
This will cause unpredictable performance impact, since each garbage collection may take an arbitrary time now.
The services should be stopped as a part of stopped flow, not on garbage collection.